server/src/main/java/com/google/coffeehouse/servlets/UpdateClubServlet.java

```diff
@@ -0,0 +1,225 @@
+   *     (and nested Book). If the request does not have the required keys, or is syntactically
+   *     incorrect, the response object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner or has an invalid ID token, the response object
+   *     will send a "403 Forbidden error"
+   * @param response the response from this method, will contain the updated Club in JSON format.
+   *     If the request object does not have a valid JSON body (as described in the request
+   *     parameter) this object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner or has an invalid ID token, the response object
+   *     will send a "403 Forbidden error"
+   * @throws IOException if an input or output error is detected when the servlet handles the request
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Club club;
+    try {
+      JsonObject requestJson = gson.fromJson(request.getReader(), JsonObject.class);
+      JsonObject updatedClubJson = requestJson.getAsJsonObject(CLUB_FIELD_NAME);
+      JsonObject updatedBookJson = updatedClubJson.getAsJsonObject(Club.CURRENT_BOOK_FIELD_NAME);
+      JsonElement updateMask = requestJson.get(UPDATE_MASK_FIELD_NAME);
+      
+      // Get attempted updates for the Club and the Book inside the Club.
+      List<String> bookMask = new ArrayList<>();
+      List<String> clubMask = new ArrayList<>();
+      if (updateMask == null) {
+        bookMask = updateableBookFields;
+        clubMask = updateableClubFields;
+      } else {
+        // Cannot use stream here because Java will complain about local variables not being final.
+        String[] elements = updateMask.getAsString().split(",");
+        for (String element : elements) {
+          boolean isBookUpdate = element.startsWith(Club.CURRENT_BOOK_FIELD_NAME + ".");
+          if (isBookUpdate) {
+            bookMask.add(element.split("\\.")[1]);
+          } else {
+            clubMask.add(element);
+          }
+        }
+      }
+
+      JsonElement clubIdElement = updatedClubJson.get(Club.CLUB_ID_FIELD_NAME);
+      if (clubIdElement == null) {
+        throw new IllegalArgumentException(String.format(NO_FIELD_ERROR, Club.CLUB_ID_FIELD_NAME));
+      }
+      club = storageHandler.fetchClubFromId(clubIdElement.getAsString());
+
+      // Determine if the user has the permissions to actually make changes on the Club.
+      JsonElement idTokenElement = requestJson.get(ID_TOKEN_FIELD_NAME);
+      if (idTokenElement == null) {
+        throw new GeneralSecurityException(INVALID_ID_TOKEN_ERROR);
+      }
+      GoogleIdToken idToken = verifier.verify(idTokenElement.getAsString());
+      if (idToken == null) {
+        throw new GeneralSecurityException(INVALID_ID_TOKEN_ERROR);
+      }
+
+      String userId = (String) idToken.getPayload().getSubject();
```

Comment:
nit: Might be good to move this logic into a shared class somewhere, so that all authenticated endpoints can use it.

Reply:
Good Idea! I've done just that.